src/mesh.cpp

```diff
@@ -172,32 +188,25 @@ void setMeshColor(scene::IMesh *mesh, const video::SColor &color)
 	}
 }
 
-void shadeMeshFaces(scene::IMesh *mesh)
+void colorizeMeshBuffer(scene::IMeshBuffer *buf, const video::SColor *buffercolor)
 {
-	if (mesh == NULL)
-		return;
-
-	u32 mc = mesh->getMeshBufferCount();
-	for (u32 j = 0; j < mc; j++) {
-		scene::IMeshBuffer *buf = mesh->getMeshBuffer(j);
-		const u32 stride = getVertexPitchFromType(buf->getVertexType());
-		u32 vertex_count = buf->getVertexCount();
-		u8 *vertices = (u8 *)buf->getVertices();
-		for (u32 i = 0; i < vertex_count; i++) {
-			video::S3DVertex *vertex = (video::S3DVertex *)(vertices + i * stride);
-			video::SColor &vc = vertex->Color;
-			if (vertex->Normal.Y < -0.5) {
-				applyFacesShading (vc, 0.447213);
-			} else if (vertex->Normal.Z > 0.5) {
-				applyFacesShading (vc, 0.670820);
-			} else if (vertex->Normal.Z < -0.5) {
-				applyFacesShading (vc, 0.670820);
-			} else if (vertex->Normal.X > 0.5) {
-				applyFacesShading (vc, 0.836660);
-			} else if (vertex->Normal.X < -0.5) {
-				applyFacesShading (vc, 0.836660);
-			}
-		}
+	const u32 stride = getVertexPitchFromType(buf->getVertexType());
+	u32 vertex_count = buf->getVertexCount();
+	u8 *vertices = (u8 *) buf->getVertices();
+	for (u32 i = 0; i < vertex_count; i++) {
+		video::S3DVertex *vertex = (video::S3DVertex *) (vertices + i * stride);
+		video::SColor &vc = vertex->Color;
+		vc = *buffercolor;
+		if (vertex->Normal.Y < -0.5)
+			applyFacesShading(vc, 0.447213);
+		else if (vertex->Normal.Z > 0.5)
+			applyFacesShading(vc, 0.670820);
+		else if (vertex->Normal.Z < -0.5)
+			applyFacesShading(vc, 0.670820);
+		else if (vertex->Normal.X > 0.5)
+			applyFacesShading(vc, 0.836660);
+		else if (vertex->Normal.X < -0.5)
+			applyFacesShading(vc, 0.836660);
```

Comment:
See https://github.com/minetest/minetest/blob/master/src/mapblock_mesh.cpp#L1167
This shading logic is differnt from that in mapblock mesh and different to lin 43 above.
See #4953 for the bug details.